base/compiler/typelattice.jl

```diff
@@ -103,9 +108,13 @@ function issubconditional(a::Conditional, b::Conditional)
     end
     return false
 end
+is_same_conditionals(a::Conditional,      b::Conditional)      = slot_id(a.var) === slot_id(b.var)
+is_same_conditionals(a::Conditional,      b::InterConditional) = slot_id(a.var) === b.slot
```

Comment:
When would we ever compare a `Conditional` to an `InterConditional`? They seem valid in exactly opposite situations, so it feels like comparing them would be a bug. Am I missing something?

Reply:
When constant propagation happens, the previous `InterConditional` result can be compared to the new `Conditional`:
https://github.com/JuliaLang/julia/blob/cec310eedb5c263d57623840b635d8f1dbf43402/base/compiler/abstractinterpretation.jl#L167

Reply:
I think we should structure the code so this does not happen. Everything should either be an interprocedurally safe type or not. Doing something like that just seems to ask for trouble. I guess the other option would be to not have `InterConditional` at all and just use `Conditional` everywhere, but if we're splitting the lattice like this, we should keep the comparisons separate.

Reply:
Ops, sorry it's just my mistake... Your insight is correct, and actually they've never been compared.
I removed those diagonal dispatch signatures.